File: 2025/day_2.py


def get_common_factors(number: int) -> list[int]:
    common_factors = []
    for i in range(1, number):
        if number % i == 0:
            common_factors.append(i)
    return common_factors


def id_is_valid(id: int) -> bool:
    str_form = str(id)
    str_len = len(str_form)
    common_factors = get_common_factors(str_len)
    for common_factor in common_factors:
        parts = set()
        index = (0, common_factor)
        while index[1] <= str_len:
            part = str_form[index[0]:index[1]]
            parts.add(part)
            index = (index[0] + common_factor, index[1] + common_factor)
            
        if len(parts) == 1:
            return False

    return True


def get_invalid_id_sum(id_ranges: str) -> int:
    ranges = id_ranges.split(",")
    invalid_sum = 0
    for id_range in ranges:
        start_index, end_index = [int(index) for index in id_range.split("-")]
        for id in range(start_index, end_index + 1):
            if not id_is_valid(id):
                print(f"Found Invalid ID = {id}")
                invalid_sum += id
    return invalid_sum
    

if __name__ == "__main__":
    id_ranges = open("2025/data/day_2_input.txt").read()
    total_invalid_sum = get_invalid_id_sum(id_ranges)
    print(f"Total Invalid Sum = {total_invalid_sum}")

File: 2025/day_2.rs
use std::fs;


fn get_common_factors(number: usize) -> Vec<usize> {
    (1..number).filter(|i| number % i == 0).map(|x| x as usize).collect()
}

fn id_is_valid(id: i64) -> bool {
    let str_form = id.to_string();
    let str_len = str_form.len();
    
    for common_factor in get_common_factors(str_len) {
        let first_chunk = &str_form[..common_factor];
        let mut all_same = true;
        
        for chunk in str_form.as_bytes().chunks(common_factor) {
            if chunk != first_chunk.as_bytes() {
                all_same = false;
                break
            }
        }
    
        if all_same {
            return false;
        }
    }
    
    true
    
}

fn get_invalid_id_sum(id_ranges: &str) -> i64 {
    let ranges: Vec<&str> = id_ranges.split(",").collect();
    let mut invalid_sum: i64 = 0;
    
    for id_range in ranges {
        let id_range = id_range.split("-").collect::<Vec<&str>>();    
        let (start_index, end_index) = (id_range.get(0).unwrap(), id_range.get(1).unwrap());
        let (start_index, end_index) = (start_index.parse::<i64>().unwrap(), end_index.parse::<i64>().unwrap());
        for id in start_index..=end_index {
            if !id_is_valid(id) {
                invalid_sum += id;
            }
        }
    }
    
    invalid_sum
}


fn main() {
    let content = fs::read_to_string("2025/data/day_2_input.txt").expect("Could not Read Input File");
    let result = get_invalid_id_sum(&content);
    println!("Total Invalid Id Sum = {result}");

}
